_internal/utils/mutate.ts

```diff
@@ -4,23 +4,40 @@ import { SWRGlobalState } from './global-state'
 import { getTimestamp } from './timestamp'
 import * as revalidateEvents from '../constants'
 import {
-  Key,
   Cache,
   MutatorCallback,
   MutatorOptions,
   GlobalState,
-  State
+  State,
+  Falsy,
+  Arguments
 } from '../types'
 
+type KeyFilter = (key?: string) => boolean
+
 export const internalMutate = async <Data>(
   ...args: [
     Cache,
-    Key,
+    KeyFilter | Arguments | Falsy,
     undefined | Data | Promise<Data | undefined> | MutatorCallback<Data>,
     undefined | boolean | MutatorOptions<Data>
   ]
-): Promise<Data | undefined> => {
-  const [cache, _key, _data, _opts] = args
+): Promise<(Data | undefined)[]> => {
+  const [cache, _keyFilter, _data, _opts] = args
+  if (!_keyFilter) return Promise.resolve([])
+  let matchedKeys: string[]
+
+  let keyFilter: KeyFilter
+  if (typeof _keyFilter !== 'function') {
+    const [serializedKey] = serialize(_keyFilter)
+    matchedKeys = [serializedKey]
+  } else {
+    keyFilter = _keyFilter as KeyFilter
+    matchedKeys = []
+    for (const _key of cache.keys()) {
+      if (keyFilter(_key)) matchedKeys.push(_key)
```

Comment:
It should pass the original key object here, but that can be a separate PR to make it easier

Reply:
I feel we need to store the original key in the cache...